File: 2023/day-2/part-1/golang/main.go
///usr/bin/true; exec /usr/bin/env go run "$0" "$@"

package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
)

func main() {
	// using bufio to read input, print result
	reader := bufio.NewReader(os.Stdin)

	thresholds := map[string]int{
		"red":   12,
		"green": 13,
		"blue":  14,
	}

	sum := 0

outer:
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			break
		}

		// Parse out game ID, which may be multiple digits
		var gameID int
		fmt.Sscanf(line, "Game %d:", &gameID)

		// Parse out each color listed
		colors := map[string]int{
			"red":   0,
			"green": 0,
			"blue":  0,
		}

		// using regex, we can parse out the colors
		regex := regexp.MustCompile(`\d+ (red|green|blue)`)
		matches := regex.FindAllStringSubmatch(line, -1)

		// loop through matches and add to color count
		for _, match := range matches {
			count := 0
			color := ""
			fmt.Sscanf(match[0], "%d %s", &count, &color)

			colors[color] = max(colors[color], count)
		}

		// if colors are less than or equal to threshold, add to sum
		for color, count := range colors {
			if count > thresholds[color] {
				continue outer
			}
		}

		sum += gameID
	}

	// print sum
	fmt.Println(sum)
}

File: 2023/day-2/part-1/rust/src/main.rs
use regex::Regex;
use std::collections::HashMap;
use std::io;

fn main() {
    // Prepare to read from stdin
    let stdin = io::stdin();
    let mut buffer = String::new();

    // Preparse running sum
    let mut sum = 0;

    // create map of thresholds
    let thresholds = HashMap::from([("red", 12), ("green", 13), ("blue", 14)]);

    // Create a regex to find counts of colors
    let re_game = Regex::new(r"Game (\d+):").unwrap();
    let re_colors = Regex::new(r"(\d+) (\w+)").unwrap();

    // Read each line
    'outer: while let Ok(n) = stdin.read_line(&mut buffer) {
        // if we have reached the end of the file, break
        if n == 0 {
            break;
        }

        // Extract game ID as a number
        let game_id = re_game.captures(&buffer).unwrap().get(1).unwrap().as_str();

        // Parse the colors and counts
        let mut counts = HashMap::new();
        for cap in re_colors.captures_iter(&buffer) {
            let count = cap.get(1).unwrap().as_str().parse::<i32>().unwrap();
            let color = cap.get(2).unwrap().as_str();

            // set counts[color] to max of count and current value
            let current = counts.entry(color).or_insert(0);
            *current = std::cmp::max(*current, count);
        }

        // Check if the counts exceed the thresholds
        for (color, count) in counts {
            if count > *thresholds.get(color).unwrap() {
                // If so, skip this game
                buffer.clear();
                continue 'outer;
            }
        }
        
        // add game ID to sum
        sum += game_id.parse::<i32>().unwrap();

        // Clear the buffer
        buffer.clear();
    }

    // Print the sum
    println!("{}", sum);
}
